### ISSUE TITLE
Do not require bitness of compiler host and target to match

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#6859]
**Original bug ID:** 6859
**Reporter:** @whitequark
**Assigned to:** @shindere
**Status:** assigned (set by @damiendoligez on 2017-03-01T13:03:01Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Version:** 4.02.2+dev / +rc1
**Category:** platform support (windows, cross-compilation, etc)
**Tags:** cross-compilation
**Monitored by:** @gasche @diml

## Bug description

Currently, it is only possible to build 32-bit binaries using a 32-bit compiler running under a 32-bit ocamlrun. This is annoying for cross-compilers, although tools like opam lessen the pain.

The chief reason for this issue is the fact that integer immediates are stored as `int`, e.g. `Const_int of int` and probably a few more places. I recall some plan for low-impact functorization of that aspect of the compiler but I could not find it on mantis, so I'm filing this issue.

**Comment by vicuna** [comment]
**Comment author:** @gasche

#170 ( https://github.com/ocaml/ocaml/pull/170 ) seems related as it keeps integers as strings during the parsing phase -- they are converted back in typecore.ml.

**Comment by vicuna** [comment]
**Comment author:** @gasche

I'm postponing this to 4.04, as no direct progress was made during the 4.03 timeline. I heard that we might see more work on cross-compilation in the medium-term future, so hopefully this will move.

**Comment by TheLortex** [comment]
Bumping this as I encountered the problem on 4.07 with the ESP32 backend (not officially supported but I tested it a lot with Mirage). Building the 32bit cross-compiler using a 64bit host compiler lead in wrong assumptions on integers. Thus I had `max_int` equal to `-1` due to incorrect optimizations.

**Comment by whitequark** [comment]
Yeah, you need the `+32bit` opam switch.

**Comment by TheLortex** [comment]
Yes that's what I currently use, so this is definitely not a blocker. It's more of a feature request to reduce the time taken to build a 32bit cross-compiler from a 64bit switch.

**Comment by whitequark** [comment]
Still an issue.

**Comment by shindere** [comment]
Does someone know how opam proceeds to deal with the issue?

It seems highly non-trivial to me,, at the moment.

I guess this will be fixed by the whole work on cross-compiling, but
it's probably not possitle to work on just fixing this directly, out of
this global cross-compiling context.

**Comment by lthls** [comment]
The opam switches for creating 32-bit compilers on 64-bit machines (with the `+32bit` suffix) simply pass some options to the `configure` script so that the compiler assumes a 32-bit architecture, and the external tools (the C compiler, linker and assembler) are called with the right options to generate 32-bit code. This works because at least on x86, 64-bit processors (and operating systems, I assume) include support for running 32-bit executables, so even executables (like ocamlrun) generated for the target can be run on the host.

There is some very limited support for correctly separating the compiler's own integers (of type `int`) from the target's integers in the `Targetint` module, and proper cross-compilation support will likely build on it, but for now it is only used in a few places.

**Comment by shindere** [comment]
OK thanks! Makes sense and things are clear!

**Comment by whitequark** [comment]
Still an issue as far as I know.
